File: ChessLab/Board/FENValidator.swift
import ChessKit

/// Validation de légalité d'un FEN, indépendamment du parseur de `ChessKit`
/// qui est volontairement permissif (il ne fait qu'une lecture structurelle).
///
/// Utilisé par la saisie FEN manuelle (étape 1) et, plus tard, par le
/// scanner d'échiquier et l'éditeur de position (étape 7) : un FEN illégal
/// ne doit **jamais** être envoyé au moteur.
enum FENValidator {

    static func isLegal(_ fen: String) -> Bool {
        errors(in: fen).isEmpty
    }

    static func errors(in fen: String) -> [String] {
        let trimmed = fen.trimmingCharacters(in: .whitespacesAndNewlines)
        let fields = trimmed.split(separator: " ").map(String.init)

        guard fields.count == 6, let position = Position(fen: trimmed) else {
            return ["FEN mal formé : 6 champs attendus (position, trait, roques, en passant, demi-coups, coups)."]
        }

        var errors: [String] = []

        let whiteKings = position.pieces.filter { $0.kind == .king && $0.color == .white }.count
        let blackKings = position.pieces.filter { $0.kind == .king && $0.color == .black }.count
        if whiteKings != 1 || blackKings != 1 {
            errors.append("Il faut exactement un roi blanc et un roi noir.")
        }

        let pawnOnEdgeRank = position.pieces.contains {
            $0.kind == .pawn && ($0.square.rank.value == 1 || $0.square.rank.value == 8)
        }
        if pawnOnEdgeRank {
            errors.append("Un pion ne peut pas être sur la 1re ou la 8e rangée.")
        }

        // Une position fraîchement chargée (aucun coup joué) ne peut avoir
        // dans son état initial qu'un `.check`/`.checkmate` portant sur le
        // camp qui N'A PAS le trait (cf. Board.updateState) : c'est
        // exactement la condition "camp adverse déjà en échec".
        if whiteKings == 1, blackKings == 1 {
            switch Board(position: position).state {
            case .check, .checkmate:
                errors.append("Le camp qui n'a pas le trait ne peut pas être déjà en échec.")
            default:
                break
            }

            // Symétrique du contrôle ci-dessus, que `Board.state` ne peut PAS
            // rendre à l'init (il n'inspecte que le camp sans le trait, voir
            // `GameOutcome.ofStartingPosition`) : un FEN où le camp AU trait
            // est déjà mat ou pat passait la validation, puis le moteur,
            // interrogé sur une position terminée, répondait `bestmove (none)`
            // — écran de jeu figé.
            if !hasLegalMove(in: position) {
                errors.append("Le camp au trait n'a aucun coup légal : la position est déjà terminée (mat ou pat).")
            }
        }

        let castling = fields[2]
        if castling != "-" {
            if castling.contains("K"), !hasRookAndKing(position, kingSquare: "e1", rookSquare: "h1", color: .white) {
                errors.append("Droit de petit roque blanc incohérent avec la position du roi/de la tour.")
            }
            if castling.contains("Q"), !hasRookAndKing(position, kingSquare: "e1", rookSquare: "a1", color: .white) {
                errors.append("Droit de grand roque blanc incohérent avec la position du roi/de la tour.")
            }
            if castling.contains("k"), !hasRookAndKing(position, kingSquare: "e8", rookSquare: "h8", color: .black) {
                errors.append("Droit de petit roque noir incohérent avec la position du roi/de la tour.")
            }
            if castling.contains("q"), !hasRookAndKing(position, kingSquare: "e8", rookSquare: "a8", color: .black) {
                errors.append("Droit de grand roque noir incohérent avec la position du roi/de la tour.")
            }
        }

        let enPassantField = fields[3]
        if enPassantField != "-", enPassantField.count == 2 {
            let epSquare = Square(enPassantField)
            let expectedPawnSquareNotation: String
            let expectedColor: Piece.Color

            switch epSquare.rank.value {
            case 3:
                expectedPawnSquareNotation = "\(epSquare.file.rawValue)4"
                expectedColor = .white
            case 6:
                expectedPawnSquareNotation = "\(epSquare.file.rawValue)5"
                expectedColor = .black
            default:
                expectedPawnSquareNotation = ""
                expectedColor = .white
                errors.append("Case en passant invalide : doit être sur la 3e ou 6e rangée.")
            }

            if !expectedPawnSquareNotation.isEmpty {
                let piece = position.piece(at: Square(expectedPawnSquareNotation))
                if piece?.kind != .pawn || piece?.color != expectedColor {
                    errors.append("Case en passant incohérente : aucun pion capturable à cet endroit.")
                }
            }
        }

        return errors
    }

    /// `Board.legalMoves(forPieceAt:)` ne consulte pas le trait : on peut
    /// donc l'interroger directement sur les pièces du camp au trait.
    private static func hasLegalMove(in position: Position) -> Bool {
        let board = Board(position: position)
        return position.pieces
            .filter { $0.color == position.sideToMove }
            .contains { !board.legalMoves(forPieceAt: $0.square).isEmpty }
    }

    private static func hasRookAndKing(
        _ position: Position,
        kingSquare: String,
        rookSquare: String,
        color: Piece.Color
    ) -> Bool {
        let king = position.piece(at: Square(kingSquare))
        let rook = position.piece(at: Square(rookSquare))
        return king?.kind == .king && king?.color == color
            && rook?.kind == .rook && rook?.color == color
    }
}

File: android/app/src/main/java/com/chesslab/editor/FenValidator.kt
package com.chesslab.editor

import chesskit.Board
import chesskit.FenParser
import chesskit.Piece
import chesskit.Position
import chesskit.Square
import com.chesslab.R

/**
 * Ce qui fait qu'une FEN décrit une position JOUABLE. Pendant de
 * `FENValidator.swift`.
 *
 * Le parseur est permissif : six jetons quelconques lui font un échiquier
 * vide sans broncher. Ici on exige ce que le jeu exige — deux rois, pas de
 * pion sur la première ou la dernière rangée, et le camp qui n'a pas le trait
 * pas déjà en échec (sinon il aurait dû jouer autre chose).
 */
object FenValidator {

    fun isLegal(fen: String): Boolean = errors(fen).isEmpty()

    /** Les défauts trouvés, en clés de ressources (`R.string`) — l'écran les met en mots. */
    fun errors(fen: String): List<Int> {
        val trimmed = fen.trim()
        if (trimmed.split(" ").filter { it.isNotEmpty() }.size != 6) return listOf(R.string.fen_error_fields)
        val position = FenParser.parse(trimmed) ?: return listOf(R.string.fen_error_fields)

        val errors = ArrayList<Int>()
        val whiteKings = position.pieces.count { it.kind == Piece.Kind.king && it.color == Piece.Color.white }
        val blackKings = position.pieces.count { it.kind == Piece.Kind.king && it.color == Piece.Color.black }
        if (whiteKings != 1 || blackKings != 1) errors += R.string.fen_error_kings

        if (position.pieces.any { it.kind == Piece.Kind.pawn && it.square.rank.value.let { r -> r == 1 || r == 8 } }) {
            errors += R.string.fen_error_pawn_rank
        }

        // Le camp qui N'A PAS le trait est-il déjà en échec ? On lui rend le
        // trait le temps de regarder : si ce plateau le dit en échec, la
        // position ne peut pas venir d'une partie.
        if (whiteKings == 1 && blackKings == 1) {
            val flipped = Position(
                position.pieces, position.sideToMove.opposite, position.legalCastlings, null, position.clock,
            )
            val state = Board(flipped).state
            if (state is Board.State.Check || state is Board.State.Checkmate) errors += R.string.fen_error_opponent_in_check

            // Le symétrique : un camp au trait sans aucun coup légal est une
            // position déjà terminée (mat ou pat). Le moteur, interrogé
            // dessus, répondrait `bestmove (none)` — écran de jeu figé.
            if (!hasLegalMove(position)) errors += R.string.fen_error_no_legal_move
        }

        val castling = trimmed.split(" ")[2]
        if (castling != "-") {
            if ('K' in castling && !hasRookAndKing(position, "e1", "h1", Piece.Color.white)) errors += R.string.fen_error_castling_wk
            if ('Q' in castling && !hasRookAndKing(position, "e1", "a1", Piece.Color.white)) errors += R.string.fen_error_castling_wq
            if ('k' in castling && !hasRookAndKing(position, "e8", "h8", Piece.Color.black)) errors += R.string.fen_error_castling_bk
            if ('q' in castling && !hasRookAndKing(position, "e8", "a8", Piece.Color.black)) errors += R.string.fen_error_castling_bq
        }

        val enPassant = trimmed.split(" ")[3]
        if (enPassant != "-" && enPassant.length == 2) {
            val file = enPassant[0]
            val (pawnSquare, pawnColor) = when (enPassant[1]) {
                '3' -> "${file}4" to Piece.Color.white
                '6' -> "${file}5" to Piece.Color.black
                else -> null to null
            }
            if (pawnSquare == null) errors += R.string.fen_error_ep_rank
            else {
                val pawn = position.piece(Square(pawnSquare))
                if (pawn?.kind != Piece.Kind.pawn || pawn.color != pawnColor) errors += R.string.fen_error_ep_pawn
            }
        }
        return errors
    }

    /** `legalMoves` ne consulte pas le trait : on interroge les pièces du camp au trait. */
    private fun hasLegalMove(position: Position): Boolean {
        val board = Board(position)
        return position.pieces.any { it.color == position.sideToMove && board.legalMoves(it.square).isNotEmpty() }
    }

    private fun hasRookAndKing(position: Position, king: String, rook: String, color: Piece.Color): Boolean {
        val k = position.piece(Square(king)); val r = position.piece(Square(rook))
        return k?.kind == Piece.Kind.king && k.color == color && r?.kind == Piece.Kind.rook && r.color == color
    }
}
